### ISSUE TITLE
Cosmos: Implement proper entity equality and contains logic for owned (and complex) types

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#34027]
When comparing a top-level, non-owned entity type to another, the EF behavior is to translate that to key comparison - we check whether the two instances are the same ("referentially identical"). However, for complex types and owned types with synthesized keys (e.g. within a JSON document), this translation is not useful - these types have no (useful) keys, and since they can only be contained by a single owner, referential comparison will always yield false (unless comparing the owned entity to itself).

As a result, we think it makes sense to perform value comparison (this is self-evident in the complex type case, but less so in the owned entity case). In the Cosmos (and general JSON) context, the most natural translation is to simply perform JSON object equality in the database. That means that `Where(b => b.ShippingAddress == b.BillingAddress)` should simply translate to `WHERE b["ShippingAddress"] = b["BillingAddress"]`, allowing the database to compare the two JSON objects.

Note that we allow users to have unmapped (unmodeled) content in JSON documents in the database. If such content exists, then the JSON object comparison may fail, as one side of the comparison may have some extra properties. This can be worked around by the user breaking down the comparison to a memberwise comparison, explicitly indicating which properties should be compared.

Note that this includes the Contains implementation as well as simple equality.

/cc @ajcvickers @AndriySvyryd

**Comment by ajcvickers** [comment]
@roji Is this still planned for EF9?

**Comment by JoasE** [comment]
Comparing 2 different owned type references (`Where(b => b.ShippingAddress == b.BillingAddress)`) will currently translate to `WHERE FALSE` because each owned property is a unique entity type. I think this is by design / correct behaviour.

Comparing to a parameter or constant (`Where(b => b.ShippingAddress == param)`) could work, but as owned entities have identities, I think it makes sense to rewrite to `Where(b => b.ShippingAddress.Id == param.Id`.

As for complex types, translating to value comparison (`b["ShippingAddress"] = b["BillingAddress"]`) will be implemented in: #37577

**Comment by roji** [comment]
> Comparing 2 different owned type references (Where(b => b.ShippingAddress == b.BillingAddress)) will currently translate to WHERE FALSE because each owned property is a unique entity type. I think this is by design / correct behaviour.

Yep. It's also why owned entity types don't make much sense for mapping nested JSON objects (or at least objects that don't have real IDs, in any case). I once again see us deprecating owned JSON mapping as we're likely to do in relational.

> Comparing to a parameter or constant (Where(b => b.ShippingAddress == param)) could work, but as owned entities have identities, I think it makes sense to rewrite to Where(b => b.ShippingAddress.Id == param.Id.

IIRC that's the right thing, but can you please confirm the relational behavior here?

**Comment by JoasE** [comment]
@roji 
To be clear, I won't be touching owned structural equality for cosmos. These are just my thoughts / context I gathered form other issues on this subject. I wanted those to add as I was in the area with #31253 , and the contents of the issue seemed inconsistent with other discussions.

As to the relational behavior; comparing an (top level) owned json entity to a const or parameter currently throws, where #36400 seems to be related as it is the same exception.
